runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/serialization/BufferingXmlStreamWriter.kt

```diff
@@ -0,0 +1,129 @@
+     * @param namespace The namespace of the tag to close (or `null` for no namespace).
+     */
+    override fun endTag(name: String, namespace: String?): XmlStreamWriter = endTag(QualifiedName(name, namespace))
+
+    private fun endTag(qName: QualifiedName): XmlStreamWriter {
+        val childWriter = tagWriterStack.removeLastOrNull()
+            ?: throw SerializationException("Unexpected end of tag while no tags are open")
+
+        if (childWriter.qName != qName) {
+            throw SerializationException("Tried to end tag $qName but expected end of ${childWriter.qName} tag")
+        }
+
+        if (tagWriterStack.isEmpty()) childWriter.write(buffer)
+        return this
+    }
+
+    /**
+     * Declares a namespace prefix for an upcoming tag (i.e., this is called before the [startTag] method 🤷‍).
+     * @param uri The URI for the namespace.
+     * @param prefix The name of the namespace which will form a prefix for tag/attribute names which utilize it.
+     */
+    override fun namespacePrefix(uri: String, prefix: String?) {
+        val attrQName = if (prefix == null) QualifiedName("xmlns") else QualifiedName(prefix, "xmlns")
+        nsAttributes[attrQName] = uri
+    }
+
+    private fun requireWriter(): LazyTagWriter {
+        if (tagWriterStack.isEmpty()) {
+            throw SerializationException("Attempted to serialize text or attribute without containing tag")
+        }
+        return tagWriterStack.last()
+    }
+
+    /**
+     * Starts this XML document by writing the XML declaration (e.g., `<?xml version="1.0"?>`).
+     * @param encoding Unused
+     * @param standalone Unused
+     */
+    override fun startDocument(encoding: String?, standalone: Boolean?) {
+        buffer.append("""<?xml version="1.0"?>""")
+        if (pretty) buffer.appendLine()
+    }
+
+    /**
+     * Starts a child tag. Note that this accumulates outstanding namespace declarations into attributes for the tag.
+     * @param name The local name of the tag.
+     * @param namespace The namespace of the tag (or `null` for no namespace).
+     */
+    override fun startTag(name: String, namespace: String?): XmlStreamWriter {
+        val currentWriter = tagWriterStack.lastOrNull()
+        val currentIndent = currentWriter?.indentLevel ?: -1
+
+        val qName = QualifiedName(name, namespace)
+        val childWriter = LazyTagWriter(pretty, currentIndent + 1, qName, nsAttributes)
+        nsAttributes.clear()
+
+        currentWriter?.childTag(childWriter)
+        tagWriterStack.addLast(childWriter)
+
+        return this
+    }
+
+    /**
+     * Adds a child text node to the most-recently opened tag.
+     */
+    override fun text(text: String): XmlStreamWriter {
+        requireWriter().text(text)
+        return this
+    }
+
+    /**
+     * Gets the text serialization for this writer. Note that this will call [endDocument] first, closing all open tags.
+     */
+    override fun toString(): String {
```

Comment:
Is this desirable?

Reply:
Is it desirable to have some code that does this? Yes, I think so.

Is it desirable that it's invoked via `toString`? Not to me, no. But it matched [the behavior of the XmlPull implementation](https://github.com/awslabs/smithy-kotlin/blob/3ac796c3bb1fcbfad4216e10c5711ba94784ff1e/runtime/serde/serde-xml/jvm/src/aws/smithy/kotlin/runtime/serde/xml/XmlStreamWriterXmlPull.kt#L79-L86) which was important for some tests (e.g., [`XmlStreamWriterTest.itWritesXMLWithAttributes`](https://github.com/awslabs/smithy-kotlin/blob/3ac796c3bb1fcbfad4216e10c5711ba94784ff1e/runtime/serde/serde-xml/common/test/aws/smithy/kotlin/runtime/serde/xml/XmlStreamWriterTest.kt#L73)) and potentially some runtime implementation. I could make a deeper effort to move any finalizing behavior to a dedicated method instead of `toString` if you like.

Reply:
> Is it desirable that it's invoked via toString? Not to me, no.
Agreed.

Up to you. I was more thinking about debugging and how easily it could mess up if you called `toString` otherwise I don't have a huge concern